WordPress/src/main/java/org/wordpress/android/util/config/AppConfig.kt

```diff
@@ -0,0 +1,61 @@
+package org.wordpress.android.util.config
+
+import org.wordpress.android.analytics.AnalyticsTracker.Stat
+import org.wordpress.android.util.analytics.AnalyticsTrackerWrapper
+import org.wordpress.android.util.config.ExperimentConfig.Variant
+import java.lang.IllegalArgumentException
+import javax.inject.Inject
+import javax.inject.Singleton
+
+@Singleton
+class AppConfig
+@Inject constructor(private val remoteConfig: RemoteConfig, private val analyticsTracker: AnalyticsTrackerWrapper) {
+    /**
+     * We need to keep the value of an already loaded feature flag to make sure the value is not changed while using the app.
+     * We should only reload the flags when the application is created.
+     */
+    private val enabledFeatures = mutableMapOf<FeatureConfig, Boolean>()
+    private val experimentValues = mutableMapOf<ExperimentConfig, String>()
+
+    /**
+     * This method initialized the config and triggers refresh of remote configuration.
+     */
+    fun refresh() {
+        remoteConfig.refresh()
+    }
+
+    /**
+     * Get the enabled state of a feature flag. If the flag is enabled in the BuildConfig file, it overrides the
+     * remote value. The correct approach is to disable a feature flag for a release version and only enable it remotely.
+     * Once the feature is ready to be fully released, we can enable the BuildConfig value.
+     * @param feature feature which we're checking remotely
+     */
+    fun isEnabled(feature: FeatureConfig): Boolean {
+        if (feature.remoteField == null) {
+            return feature.buildConfigValue
+        }
+        return enabledFeatures.getOrPut(feature) {
+            val loadedValue = feature.buildConfigValue || remoteConfig.isEnabled(feature.remoteField)
```

Comment:
If the buildValue is `false` and the remoteValue is `true`, the feature will be enabled. So we can theoretically accidentally enable the feature before it's released (before buildValue == true), right? How would you feel about using `&&` instead to make sure we never release it prematurely?

Moreover, this solution basically ignores the remoteValue if the buildValue == true -> we can't "unrelease" a feature when we find a bug. Wdyt?

Update: AFAIU the idea is to remove the buildValue altogether when we are about to release the feature, right? I'm a bit worried this will be a source of mistakes as developers will just set the buildValue == true, instead of removing it.

Reply:
I'm trying to understand this as well. The BuildConfig value overrides the remote value, but only when the BuildConfig value == true, is that correct? What is the default value of the flag when instances in which we can't connect to FB? (Rare, but possible)

Reply:
Hey, this is intentional. I think I need to explain the `BuildConfig` value. When you start developing a feature, you set the `BuildConfig` to false and you don't create a `remote_field` in Firebase. This makes the feature (in that state) unreleasable. Once the feature is ready for an A/B test, you can create the remote field and set it here. At the same time you want to be able to enable the flag for debug/testers (this is the reason why I've used the `||`). Basically you'd set the BuildConfig value for `debug` to true, for `release` to false and this will enable you to partially release your feature which is not fully released. Once your development is ready, you can set the build config value to `true` for all the flavours. At this time you can't disable the feature remotely. 

> I'm trying to understand this as well. The BuildConfig value overrides the remote value, but only when the BuildConfig value == true, is that correct? What is the default value of the flag when instances in which we can't connect to FB? (Rare, but possible)

The default value is defined in the `remote_config_defaults` file. If the flag is not there, it defaults to `false`. When the flag is not present, you still want to be able to enable the flag locally. That's why I'm using the `||` here. With `&&` you'd always need to have the remote value set to enable the feature.

Reply:
> Once your development is done, you can set the build config value to true for all the flavours. At this time you can't disable the feature remotely.

Thanks for the reply. I understand and agree with most things you said. However, this part still confuses me. Isn't one of the main points of feature flags, the ability to disable the feature when something goes wrong?

Reply:
> However, this part still confuses me. Isn't one of the main points of feature flags, the ability to disable the feature when something goes wrong?

I think it's very rare that you want to revert a feature remotely that's fully released to everyone. The flow should be that you use a feature flag (and keep the BuildConfig value to `false`) and you enable the feature in Firebase to a small subset of users (5%?). If there are no crashes, you can increase it to a higher number. Ultimately you end up with 100% of users having the feature enabled remotely. This doesn't mean the feature is **done**. Nothing forces you to set the `BuildConfig` value to `true` for release. We can actually omit it from the steps and only leave it for debug/testers. We definitely need a way to set the FF to true for a given build. 

I think one additional step should be adding a way to set feature flags manually in the settings (for debug/beta builds). If we do this, we no longer need the BuildConfig values.

Reply:
ohh I see. Thanks for the clarification!

>We definitely need a way to set the FF to true for a given build.

This is something I have been [dealing with last week](https://github.com/wordpress-mobile/WordPress-Android/pull/12050). Enabling an unfinished feature on a specific build flavors, might not be enough for us. I believe we'll need to enable it just for certain users.

I like the idea of using a debug-only screen where every developer can enable/disable the features they want. We could implement this during the hack week. Wdyt?

Reply:
> I like the idea of using a debug-only screen where every developer can enable/disable the features they want. We could implement this during the hack week. Wdyt?

Let's do this if we have time. We can also have it generated with an annotation processor :-)